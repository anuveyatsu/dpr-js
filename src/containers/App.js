import React, {PropTypes} from "react";
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

class App extends React.Component {

    constructor(props, context) {
        super(props, context);

    }

    render() {
		return (
			<div>
				{this.props.children}
			</div>
		);
    }
}
App.propTypes = {
	children: PropTypes.element
};

function mapStateToProps(state) {
	return {
		state: state
	};
}

// function mapDispatchToProps(dispatch) {
// 	return {
// 		authActions: bindActionCreators(authActions, dispatch)
// 	};
// }

export default connect(mapStateToProps)(App);
