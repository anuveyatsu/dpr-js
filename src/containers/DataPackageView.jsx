import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import _ from 'lodash'
const Datapackage = require('datapackage').Datapackage


import * as actions from '../actions/datapackageActions'
import * as dputils from '../../src/utils/datapackage'
import * as viewutils from '../utils/view'
import HandsOnTable from '../components/dataPackageView/HandsOnTable'
import PlotlyChart from '../components/dataPackageView/PlotlyChart'
import VegaLiteChart from '../components/dataPackageView/VegaLiteChart'

export class DataPackageView extends React.Component {
  constructor(props) {
    super(props)
    // TODO: what is the point of state? Why not just use props?
    this.state = {
      // we stub some basic fields to ensure render works ...
      dataPackage: Object.assign({resources: [], views: []}, this.props.dataPackage),
      dataPackageUrl: this.props.dataPackageUrl
    }
  }

  componentDidMount() {
    // if current dataPackage is empty let's load it ...
    // TODO: testing existence of name property is hacky way to check data
    // package exists
    if (!this.state.dataPackage.name) {
      dputils.fetchDataPackageAndData(this.state.dataPackageUrl).then(dpObj => {
        this.setState({
          dataPackage: dpObj.descriptor,
        })
      });
    }
  }

  render() {
    let dp = this.state.dataPackage
    let viewComponents = dp.views.map((view, idx) => {
      // first let's fix up recline views ...
      if (view.type == 'Graph') { // it's a recline view
        view = viewutils.convertReclineToSimple(view)
      }
      let compiledView = viewutils.compileView(view, dp)
      switch (view.specType) {
        case 'simple': // convert to plotly then render
          let spec = viewutils.simpleToPlotly(compiledView)
          return <PlotlyChart data={spec.data} layout={spec.layout} idx={idx} />
      }
    })
    let resourceDisplayComponents = dp.resources.map((resource, idx) => {
      let compiledViewSpec = {
        resources: [resource],
        specType: 'handsontable'
      }
      let spec = viewutils.handsOnTableToHandsOnTable(compiledViewSpec)
      return <HandsOnTable spec={spec} idx={idx} key={idx} />
    });
    return <div>{viewComponents.concat(resourceDisplayComponents)}</div>
  }
}

function mapStateToProps(state, ownProps) {
  const { descriptor, resources, readme } = state.dpr

  return {
    descriptor
    , resources
    , readme
  }
}

function mapDispatchToProps(dispatch) {
  return {
    dataPackageActions: bindActionCreators(actions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(DataPackageView)
