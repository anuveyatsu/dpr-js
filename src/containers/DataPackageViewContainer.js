import React, {PropTypes} from "react";
import DataPackagePanel from "../components/dataViewPanel/DataPackagePanel";
import { connect } from 'react-redux';
import {bindActionCreators} from "redux";
import * as actions from '../actions/datapackageActions';

//This container component listens to updates in datapackage and resources from
//the Redux Store. It then generates either Plotly or Vega-lite spec and renders
//appropriate chart.
class DataPackageViewContainer extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      specs:{}
    };
  }

  componentDidMount(){
    let url = "https://raw.githubusercontent.com/anuveyatsu/test_data/master/datapackage.json";
    this.props.dataPackageActions.getDataPackage(url);
  }

  async componentWillReceiveProps(nextProps) {
    if(nextProps.datapackage.resources) {
      //check if resources are received by comparing descriptor's resources and
      //received data length
      if(nextProps.datapackage.resources.length === nextProps.resources.length) {
        let specs = await this.getSpecsFromNextProps(nextProps.datapackage, nextProps.resources);
        this.setState({specs: specs});
      }
    }
  }

  //Takes data and view, then generates vega-lite specific spec.
  generateVegaLiteSpec(data, view) {
    let vlSpec = {
      "width": 900,
      "height": 400,
      "data": {"values": []},
      "layers": []
    };
    let headers = data[0];
    vlSpec.data.values = data.slice(1).map(values => {
      return headers.reduce((o, k, i) => {
        o[k] = values[i];
        return o;
      }, {});
    });
    for(let i=0; i<view.state.series.length; i++) {
      let layer = {
        "mark": "line",
        "encoding": {
          "x": {"field": "", "type": "temporal"},
          "y": {"field": "", "type": "quantitative"}
        }
      };
      layer.encoding.x.field = view.state.group;
      layer.encoding.y.field = view.state.series[i];
      vlSpec.layers.push(layer);
    }
    return vlSpec;
  }

  //Takes a single resource and descriptor, then converts resource into Plotly
  //specific format and generates plotlySpec.
  generatePlotlySpec(data, dp, j) {
    let plotlySpec = {};
    let dataset = [];
    let group = dp.views[j].state.group;
    let series = dp.views[j].state.series;
    let xIndex;
    let yIndex = [];
    data[0].forEach((header, index) => {
      if(header === group) {xIndex = index;}
      series.forEach(serie => {
        if(header === serie) {yIndex.push(index);}
      });
    });
    for (let i = 0; i < series.length; i++) {
      dataset.push({x: [], y: [], mode: "lines", name: series[i]});
      dataset[i].x = data.slice(1).map(row => row[xIndex]);
      dataset[i].y = data.slice(1).map(row => row[yIndex[i]]);
    }

    let layout = {
      "xaxis": {
        "title": dp.views[j].state.group
      }
    };
    plotlySpec.data = dataset;
    plotlySpec.layout = layout;
    return plotlySpec;
  }

  //Takes a single resource and returns Handsontable spec
  generateHandsontableSpec(data) {
    return {
      data: data.slice(1), //excluding headers
      colHeaders: data[0], //selecting headers
      readOnly: true,
      width: 1136,
      height: function () {
        if (data.length > 16) {
          return 432;
        }
      },
      colWidths: 47,
      rowWidth: 27,
      stretchH: 'all',
      columnSorting: true,
      search: true
    };
  }

  async getSpecsFromNextProps(datapackage, resources){
    let specs = {};

    for (let i=0; i<resources.length; i++) {
      let graphType;
      let spec = {};
      if(datapackage.views) {
        graphType = datapackage.views[i].type;
      }
      spec['htSpec'] = this.generateHandsontableSpec(resources[i]);
      if(graphType === "vega-lite") {
        let vlSpec = this.generateVegaLiteSpec(resources[i], datapackage.views[i]);
        spec['graphType'] = graphType;
        spec['vlSpec'] = vlSpec;
      } else {
        let plotlySpec = this.generatePlotlySpec(resources[i], datapackage, i);
        spec['graphType'] = graphType;
        spec['plotlySpec'] = plotlySpec;
      }
      specs[i] = spec;
    }
    return specs;
  }

  render() {
    return (
      <DataPackagePanel specs={this.state.specs}/>
    );
  }
}

DataPackageViewContainer.propTypes = {
  dataPackageActions: PropTypes.object
};

function mapStateToProps (state, ownProps) {
  const { datapackage, resources } = state.dpr;

  return {
    datapackage: datapackage,
    resources: resources
  };
}

function mapDispatchToProps(dispatch) {
  return {
    dataPackageActions: bindActionCreators(actions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(DataPackageViewContainer);
