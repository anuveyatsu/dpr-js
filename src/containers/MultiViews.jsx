import urllib from 'url'

import React, { PropTypes } from 'react'
import {CopyToClipboard} from 'react-copy-to-clipboard'

import * as dprender from 'datapackage-render'
import PlotlyChart from '../components/dataPackageView/PlotlyChart'
import VegaChart from '../components/dataPackageView/VegaChart'
import HandsOnTable from '../components/dataPackageView/HandsOnTable'

export class MultiViews extends React.Component {
  constructor(props) {
    super(props)
    // TODO: what is the point of state? Why not just use props?
    this.state = {
      // we stub some basic fields to ensure render works ...
      dataPackage: this.props.dataPackage
    }
  }


  render() {
    let dp = this.state.dataPackage
    let viewComponents
    if(dp.views) {
      viewComponents = dp.views.map((view, idx) => {
        // check if the view is not a preview
        if (!view.datahub || !(view.datahub.type === 'preview')) {
          // first let's fix up recline views ...
          if (view.type == 'Graph') { // it's a recline view
            view = dprender.convertReclineToSimple(view)
          }
          let compiledView = dprender.compileView(view, dp)
          let readyView
          switch (view.specType) {
            case 'simple': // convert to plotly then render
              let spec = {}
              if(compiledView.resources[0]._values) {
                spec = dprender.simpleToPlotly(compiledView)
              }
              readyView = <PlotlyChart data={spec.data} layout={spec.layout} idx={idx} key={idx} />
              break
            case 'vega': // render VegaChart
              let vegaSpec = dprender.vegaToVega(compiledView)
              readyView = <VegaChart spec={vegaSpec} idx={idx} key={idx} />
              break
            case 'table': // render handsontable
              let htSpec = dprender.handsOnTableToHandsOnTable(compiledView)
              readyView = <HandsOnTable spec={htSpec} idx={idx + 'v'} key={idx} />
              break
          }
          let baseUrl = window.location.href
          baseUrl = baseUrl.endsWith('/') ? baseUrl : baseUrl + '/'
          const viewPath = `view/${idx}`
          const sharedUrl = urllib.resolve(baseUrl, viewPath)
          const iframe = `<iframe src="${sharedUrl}" width="100%" height="100%" frameborder="0"></iframe>`
          const embedableId = `embed${idx}`
          return (
            <div>
              {readyView}
              <div className="share-and-embed">
                <span className="copy-text">Share:</span>
                <input value={sharedUrl} className="copy-input" />
                  <CopyToClipboard text={sharedUrl}
                    onCopy={() => console.log('Copied to clipboard')}>
                    <button className="copy-button">Copy</button>
                  </CopyToClipboard>

                <span className="copy-text">Embed:</span>
                <input value={iframe} className="copy-input" />
                <CopyToClipboard text={iframe}
                  onCopy={() => console.log('Copied to clipboard')}>
                  <button className="copy-button">Copy</button>
                </CopyToClipboard>
              </div>
            </div>
          )
        }
      })
    }
    return <div>{viewComponents}</div>
  }
}

export default MultiViews
