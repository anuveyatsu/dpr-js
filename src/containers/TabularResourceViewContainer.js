import React from "react"
import HandsOnTable from "../components/handsontable"
//connect redux:
import { connect } from 'react-redux'
import * as actions from '../actions/datapackageActions'

export class TabularResourceViewContainer extends React.Component {

  constructor(props) {
    super(props)

    this.state = {
      spec: []
    }
  }

  buildHandsontableSpec(data) {
    let spec = {
      data: data.slice(1),
      colHeaders: data[0],
      readOnly: true,
      width: 1136,
      height: function(){
        if (data.length > 16) {return 432;}
      },
      colWidths: 47,
      rowWidth: 27,
      stretchH: 'all',
      columnSorting: true,
      search: true
    }
    return spec
  }

  async componentWillReceiveProps(nextProps) {
    if(nextProps.datapackage) {
      if(nextProps.datapackage.resources.length == nextProps.resources.length) {
        let spec = await this.buildHandsontableSpec(nextProps.resources[0][this.props.idx])
        this.setState({
          spec: spec
        })
      }
    }
  }

  render() {
    return (
      <HandsOnTable spec={this.state.spec} />
    )
  }
}

const mapStateToProps = (state) => {
  const { datapackage, resources } = state

  return {
    datapackage: datapackage,
    resources: resources
  }
}

export default connect(mapStateToProps)(TabularResourceViewContainer)
