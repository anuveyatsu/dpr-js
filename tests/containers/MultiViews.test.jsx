import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { MultiViews } from '../../src/containers/MultiViews'

const mockDescriptor = {
  name: 'demo-package'
  , resources: [
    {
      name: 'demo-resource'
      , path: 'data/demo-resource.csv'
      , format: 'csv'
      , mediatype: 'text/csv'
      , schema: {
        fields: [
          {
            name: 'Date'
            , type: 'date'
            , description: ''
          }
          , {
            name: 'Open'
            , type: 'number'
            , description: ''
          }
          , {
            name: 'High'
            , type: 'number'
            , description: ''
          }
        ]
        , primaryKey: 'Date'
      }
      , _values: [
        ['2014-01-01', 14.32, 14.59]
        , ['2014-01-02', 14.06, 14.22]
        , ['2014-01-05', 13.41, 14.00]
      ]
    }
  ]
  , views: [
    {
      id: 'Graph'
      , type: 'Graph'
      , state: {
        graphType: 'lines'
        , group: 'Date'
        , series: ['High']
      }
    }
    , {
      name: 'graph'
      , specType: 'simple'
      , spec: {
        type: 'line'
        , group: 'Date'
        , series: ['High']
      }
    }
    , {
      name: 'vega-graph'
      , specType: 'vega'
      , spec: {
        width: 400
        , height: 200
        , scales: []
        , axes: []
        , marks: []
      }
    }
    , {
      name: 'table'
      , specType: 'table'
    }
  ]
}

describe('MultiViews Container', () => {
  it('should render 3 charts + 1 table', () => {
    const wrapper = shallow(<MultiViews dataPackage={mockDescriptor} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should render 2 PlotlyChart components and 1 VegaChart component', () => {
    mockDescriptor.views.splice(-1, 1)
    const wrapper = shallow(<MultiViews dataPackage={mockDescriptor} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should render empty PlotlyChart / VegaChart if there is no data yet', () => {
    delete mockDescriptor.resources[0]._values
    const wrapper = shallow(<MultiViews dataPackage={mockDescriptor} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should NOT render PlotlyChart / VegaChart if there is no views given', () => {
    delete mockDescriptor.views
    const wrapper = shallow(<MultiViews dataPackage={mockDescriptor} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
