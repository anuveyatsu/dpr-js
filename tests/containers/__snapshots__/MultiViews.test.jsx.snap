exports[`MultiViews Container should NOT render PlotlyChart if there is no views given 1`] = `<div />`;

exports[`MultiViews Container should render PlotlyChart component 1`] = `
<div>
  <PlotlyChart
    data={
      Array [
        Object {
          "line": Object {
            "shape": "spline",
            "width": 2,
          },
          "mode": "lines",
          "name": "High",
          "type": "scatter",
          "x": Array [
            "2014-01-01",
            "2014-01-02",
            "2014-01-05",
          ],
          "y": Array [
            14.59,
            14.22,
            14,
          ],
        },
      ]
    }
    idx={0}
    layout={
      Object {
        "height": 450,
        "title": "demo-resource",
        "xaxis": Object {
          "title": "Date",
        },
        "yaxis": Object {
          "title": "High",
        },
      }
    } />
  <PlotlyChart
    data={
      Array [
        Object {
          "line": Object {
            "shape": "spline",
            "width": 2,
          },
          "mode": "lines",
          "name": "High",
          "type": "scatter",
          "x": Array [
            "2014-01-01",
            "2014-01-02",
            "2014-01-05",
          ],
          "y": Array [
            14.59,
            14.22,
            14,
          ],
        },
      ]
    }
    idx={1}
    layout={
      Object {
        "height": 450,
        "title": "demo-resource",
        "xaxis": Object {
          "title": "Date",
        },
        "yaxis": Object {
          "title": "High",
        },
      }
    } />
</div>
`;

exports[`MultiViews Container should render empty PlotlyChart if there is no data yet 1`] = `
<div>
  <PlotlyChart
    idx={0} />
  <PlotlyChart
    idx={1} />
</div>
`;
